feat: add `random/base/tinymt32`

abdelrahman04 opened the pull request:
Resolves #200 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Draft PR for Adding TinyMT32, a pseudo random number generator

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #200  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

abdelrahman04 committed:
feat(prng): add tinymt32 pseudorandom number generator

abdelrahman04 committed:
feat(prng): add tinymt32 fixtures for testing

abdelrahman04 committed:
feat(prng): corrected license year

abdelrahman04 commented:
@kgryte Hello Athan,
I believe the implementation is all set
just missing the correct initialization from a seed array, it isn't done in the original author's implementation so I will check more on what to do in that case,
I made sure to have the tinymt32 match the mt19937 format in functions and descriptions mostly
for the normalized function, I followed your approach of the 2^53 precision instead of the authors way ( just multiplying by 1 / MAX Unsigned Int range).
Also, I know the linting has some issues, I just want to make sure everything is good while finalizing that and would like to know if you know of a vscode extension to point out the linting errors
Thanks

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/lib/factory.js:
```diff
@@ -0,0 +1,556 @@
+    return null;
+}
+
+function multiplyUnsigned(a, b) {
```
Not clear why you are not using `number/uint32/base/mul`, rather than including directly here.

abdelrahman04 replied:
You are right, I am just not that familiar with most of the number/ usage yet
Will change.

kgryte commented:
Yes, this PR follows the general format that we are looking for. The package needs significant clean-up, however. I suggest ensuring that you have followed the project development guide as there it tells you that you should have setup EditorConfig, etc.

For some recommended VSCode configurations, see https://github.com/stdlib-js/stdlib/tree/develop/docs/editors/vscode. It is possible, however, that that guide may be slightly out-of-date.

abdelrahman04 committed:
feat(lint): corrected linting for new package tinymt32

abdelrahman04 committed:
Merge branch 'stdlib-js:develop' into tiny-random

abdelrahman04 committed:
feat(lint): corrected linting for markdown and javascript files for the new package tinymt32

abdelrahman04 committed:
feat(lint): corrected linting for c files for the new package tinymt32

abdelrahman04 committed:
feat(lint): corrected linting for editorconfig for the new package tinymt32

abdelrahman04 committed:
feat(lint): corrected linting for c files

abdelrahman04 committed:
feat(lint): changed repl format to follow docs

abdelrahman04 committed:
feat(lint): changed package.json to remove unwanted files

abdelrahman04 committed:
feat(lint): corrected example.c to match linter

abdelrahman04 commented:
@kgryte 
Hello Athan,
I think all linting has been made correctly, the package is almost ready and waiting for review.
1- for the `example/c/example.c` I had to rewrite the tinymt32.h again since it didn't capture it outside from it's directory no matter what, I tried to look at mt19937 of handling it but found that the package is messy and doesn't even compile for the c code any more so I think it needs cleaning that can be addressed in an issue.
2- for the commits, I had to commit multiple times which were redundant too, should I create another PR with a clean branch that only has 1 commit?
Thanks for your review.

kgryte commented:
> I tried to look at mt19937 of handling it but found that the package is messy and doesn't even compile for the c code any more so I think it needs cleaning that can be addressed in an issue.

I am not sure I follow. Here is what I see when compiling the MT19937 C example:

```
$ make examples-c EXAMPLES_FILTER=".*/random/base/mt19937/.*"

Running example: /stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/examples/c/example.c
Resolving package path...
Package path: /stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937
Resolving package manifest...
Successfully resolved package manifest.
Resolving include directories...
Resolving source files...
Resolving libraries...
Resolving library paths...
Compiling example...
Successfully compiled example.
Success!

seed = 12345
name = mt19937
min = 0
max = 4294967295

Pseudorandom integers...
1789368711
3146859322
43676229
3522623596
3544234957
3448207591
1282648386
3672791226
1582316135
4001984784

min (normalized) = 0.0000000000000000
max (normalized) = 0.9999999999999999

Pseudorandom doubles...
0.1936613490450743
0.5660081687288613
0.1616878239293682
0.1242668842835302
0.4329362680099159
0.5620784880758429
0.1743435607237318
0.5532210855693298
0.3549013863365987
0.9580647850995486
```

It is quite possible that you are not compiling things correctly.

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/examples/c/tinymt32.h:
This file shouldn't be here. Please study how to compile packages using project tooling.

abdelrahman04 replied:
When I followed mt19937 way of compiling, the lining was failing, should proceed with needed conventions without regarding exact linting?
I tried to check the linting for mt19937 and found that it was failing for c files, even though functionally it was alright, that's what I meant in my comment above
I can make a pr with some subtle changes in mt19937 to let you see it

kgryte replied:
Here is what I see when linting MT19937:

```bash
$ make lint-c SOURCES_FILTER=".*/random/base/mt19937/.*" TESTS_FIXTURES_FILTER=".*/random/base/mt19937/.*" EXAMPLES_FILTER=".*/random/base/mt19937/.*" BENCHMARKS_FILTER=".*/random/base/mt19937/.*"

Linting file: stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/src/main.c
(information) Couldn't find path given by -I 'stdlib/stdlib/lib/node_modules/@stdlib/ndarray/base/unary-reduce/include/'
nofile:0:0: information: Active checkers: 114/835 (use --checkers-report=<filename> to see details) [checkersReport]


Linting file: stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/examples/c/example.c
(information) Couldn't find path given by -I 'stdlib/stdlib/lib/node_modules/@stdlib/ndarray/base/unary-reduce/include/'
nofile:0:0: information: Active checkers: 115/835 (use --checkers-report=<filename> to see details) [checkersReport]


Linting file: stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/benchmark/c/benchmark.c
(information) Couldn't find path given by -I 'stdlib/stdlib/lib/node_modules/@stdlib/ndarray/base/unary-reduce/include/'
nofile:0:0: information: Active checkers: 114/835 (use --checkers-report=<filename> to see details) [checkersReport]


Linting file: stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/test/fixtures/c/runner.c
(information) Couldn't find path given by -I 'stdlib/stdlib/lib/node_modules/@stdlib/ndarray/base/unary-reduce/include/'
stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/test/fixtures/c/runner.c:151:46: style: Parameter 'x' can be declared as pointer to const [constParameterPointer]
void ui32_to_f64( double *out, unsigned int *x, unsigned int len ) {
                                             ^
stdlib/stdlib/lib/node_modules/@stdlib/random/base/mt19937/test/fixtures/c/runner.c:166:46: style: Parameter 'x' can be declared as pointer to const [constParameterPointer]
void f64_to_ui32( unsigned int *out, double *x, unsigned int len ) {
                                             ^
nofile:0:0: information: Active checkers: 114/835 (use --checkers-report=<filename> to see details) [checkersReport]

make[1]: *** [cppcheck-tests-fixtures] Error 1
make: *** [lint-c-tests-fixtures] Error 2
```

The only lint errors are in the fixture runner and these are very minor.

abdelrahman04 replied:
solved

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/examples/c/tinymt32.c:
This file should be removed.

abdelrahman04 replied:
Done

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/examples/c/Makefile:
This is not our Makefile and should be replaced with the Makefile used in other PRNGs.

abdelrahman04 replied:
Corrected

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/examples/c/example.c:
```diff
@@ -0,0 +1,64 @@
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include "tinymt32.h"
```
This is wrong. Everything should be based on stdlib paths.

kgryte replied:
Again, you should be replicating what you see in `random/base/lcg` and `random/base/mt19937` exactly. If you cannot get things to compile, it is much more likely that you are doing things incorrectly, than that our packages are broken.

abdelrahman04 replied:
Done with the exact same format (imports)

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/test/fixtures/c/Makefile:
This is not our Makefile convention. We use project tooling for compiling test fixtures. See the root `/tools/make` directory.

kgryte replied:
This file should be replaced with https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/random/base/mt19937/test/fixtures/c/Makefile

abdelrahman04 replied:
Done

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/test/fixtures/c/tinymt32.h:
Remove this file.

abdelrahman04 replied:
Done

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/include/stdlib/random/base/tinymt32.h:
```diff
@@ -0,0 +1,79 @@
+	uint32_t *seed;
+	int64_t seed_length;
+	uint32_t state[4];
+	uint32_t mat1;
+	uint32_t mat2;
+	uint32_t tmat;
```
Why are these fields necessary? They do not appear to be used.

abdelrahman04 replied:
they are used in the `tinymt32/src/main.c`
check the latest commit, their uses should be clear now,
a coefficient used to get next state and in the initializing of states.

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/include/stdlib/random/base/tinymt32.h:
```diff
@@ -0,0 +1,79 @@
+/**
+* Returns a pointer to a dynamically allocated PRNG.
+*/
+struct BasePRNGObject * stdlib_base_random_tinymt32_allocate( uint32_t seed, int64_t seed_length ) __attribute__((unused));
```
Why do you have `__attribute__((unused))` here?

abdelrahman04 replied:
when linting locally, it always mentioned unused even though it's used, had to silence it
will investigate to remove if safely

abdelrahman04 replied:
Corrected and removed

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/benchmark/c/benchmark.c:
This file needs to be rewritten to follow https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/random/base/mt19937/benchmark/c/benchmark.c.

abdelrahman04 replied:
Done

kgryte commented on lib/node_modules/@stdlib/random/base/tinymt32/LICENSE:
```diff
@@ -0,0 +1,216 @@
+implementations from the following external libraries, which have their own
+licenses:
+
+* MT19937 <http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/CODES/
```
This isn't the correct license for TinyMT32.

abdelrahman04 replied:
https://github.com/MersenneTwister-Lab/TinyMT/blob/master/LICENSE.txt
the math.sci link references their main distribution, and the above repo references their main implementation that has been moved, should I use this license?

kgryte replied:
What is in https://www.math.sci.hiroshima-u.ac.jp/m-mat/MT/TINYMT/LICENSE.txt is essentially the same as https://github.com/MersenneTwister-Lab/TinyMT/blob/master/LICENSE.txt. The latter simply updates the copyright date. As such, using the latter is fine.

abdelrahman04 replied:
Corrected.

abdelrahman04 committed:
feat(prng): corrected project tools to enhance structure

abdelrahman04 committed:
feat(lint): correcting linting for editorconfig

abdelrahman04 committed:
feat(lint): correcting linting for c files

abdelrahman04 commented:
@kgryte 
Hello Athan,
I Corrected all reviewed issues, ran the following
```
make examples EXAMPLES_FILTER=".*/random/base/tinymt32/.*"
make examples-c EXAMPLES_FILTER=".*/random/base/tinymt32/.*"
make benchmark BENCHMARKS_FILTER=".*/random/base/tinymt32/.*"
make benchmark-c BENCHMARKS_FILTER=".*/random/base/tinymt32/.*"
make test TESTS_FILTER=".*/random/base/tinymt32/.*"
```
I don't know if there is anything else that need to be checked but I think the PR is ready for merging, should I do another branch with 1 clean commit?
Thanks

abdelrahman04 committed:
feat(lint): correcting linting for main c files

kgryte commented:
> should I do another branch with 1 clean commit?

No, as we perform squash merging.

abdelrahman04 committed:
Merge branch 'stdlib-js:develop' into tiny-random

abdelrahman04 commented:
> No, as we perform squash merging.

@kgryte 
Is it ready for merging, or does it require further change?
Thanks

abdelrahman04 commented:
@kgryte 
Should I create a new non-draft PR?
Thanks

kgryte committed:
bench: update benchmarks to actually test factory function creation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix missing `normalized` benchmark

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: fix description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix value limits

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update markup

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
